### PR TITLE
stats: Add quic connection close error code histogram.

### LINKED ISSUE #10102: quiche: implement QUIC stats API
source/extensions/quic_listeners/quiche/platform/quic_server_stats_impl.h has no-op implementation for QUIC stats interface, i.e. QUIC_SERVER_HISTOGRAM_ENUM_IMPL. We need to implement those with Envoy stats.

### PR DESCRIPTION
**Pull request by RenjieTang**
Signed-off-by: Renjie Tang <renjietang@google.com>

Commit Message: Add connection close error code histograms for quic. This is the one of the most important metrics for quic experimentations.
Additional Description:
Risk Level: low
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
Fixes #10102

### CODE REVIEW

**Comment by RenjieTang** [comment]
/assign @RyanTheOptimist
Ryan would you mind taking a first-pass review?

File: test/common/quic/envoy_quic_client_session_test.cc [hunk 1]
```diff
@@ -133,15 +133,17 @@ class EnvoyQuicClientSessionTest : public testing::TestWithParam<bool> {
...
+                                            write_current_, nullptr, nullptr,
+                                            &self_connection_close_error_, nullptr});
     EXPECT_EQ(&read_total_, &quic_connection_->connectionStats().read_total_);
   }
 
   void TearDown() override {
     if (quic_connection_->connected()) {
       EXPECT_CALL(*quic_connection_,
                   SendConnectionClosePacket(quic::QUIC_NO_ERROR, _, "Closed by application"));
+      EXPECT_CALL(self_connection_close_error_, recordValue(0));
```
Comment (RyanTheOptimist) [inline test/common/quic/envoy_quic_client_session_test.cc:146]: Are there any places where we might be able to EXPECT_CALL a non-zero value? We may well not have any such tests at the envoy level?

**Review by RyanTheOptimist — APPROVED** [review]
Nice! That's a lot of files to change just to add two histograms :)

File: test/common/quic/envoy_quic_client_session_test.cc (hunk 1 above)
Comment (RenjieTang) [reply test/common/quic/envoy_quic_client_session_test.cc:146]: Right now we don't. But if more tests are added in this file in the future, I expect we will have recordValue() with non-zero parameter.

File: include/envoy/network/connection.h [hunk 2]
```diff
@@ -96,6 +96,10 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
     Stats::Counter* bind_errors_;
     // Optional counter. Delayed close timeouts will not be tracked if this is nullptr.
     Stats::Counter* delayed_close_timeouts_;
+    // Optional histogram, used to track quic self-initiated connection close error if not nullptr.
+    Stats::Histogram* quic_connection_self_close_error_;
```
Comment (ggreenway) [inline include/envoy/network/connection.h:100]: I think these stats should be on a transport_socket or similar, because they don't apply to non-quic connections.

**Review by ggreenway — CHANGES_REQUESTED** [review]
I'm not sure a histogram is the correct type for error codes; doing things like percentiles on error codes doesn't make sense. cc @jmarantz 

The increase in memory use, as noted in the changes to the tests, seems pretty steep, given that it applies to non-quic configurations as well. 

/wait

**Comment by RenjieTang** [comment]
Thanks for the review! My intention for this change is to somehow record QUIC connection close error codes so that users can analyze their distributions. I have 2 questions:
1. Does declaring HISTOGRAM in upstream.h invoke memory allocation right away even if no QUIC connection will be used and no value will be recorded?
2. If histogram is too unwieldy, what other approaches are available in Envoy?

**Comment by ggreenway** [comment]
> Thanks for the review! My intention for this change is to somehow record QUIC connection close error codes so that users can analyze their distributions. I have 2 questions:
> 
> 1. Does declaring HISTOGRAM in upstream.h invoke memory allocation right away even if no QUIC connection will be used and no value will be recorded?

Yes, this causes the stat to be allocated before any connections have been made (when the config is loaded).

> 2. If histogram is too unwieldy, what other approaches are available in Envoy?

The most similar example I can think of is the grpc_stats filter, which records metrics for how many times each grpc-status is encountered. It dynamically creates stats for each value the first time each one is encountered, using a COUNTER for each one.

There are similar stats for some http codes. There are some bucket stats (`2xx`, `3xx`, `4xx`, `5xx`), as well as some counters for specific errors like 502.

File: include/envoy/network/connection.h (hunk 2 above)
Comment (danzh2010) [reply include/envoy/network/connection.h:100]: source/common/http/http3/codec_stats.h might be a better place for them.

**Review by danzh2010 — COMMENTED** [review]
> I'm not sure a histogram is the correct type for error codes; doing things like percentiles on error codes doesn't make sense. cc @jmarantz

Agree that Envoy histogram might not be a proper way to aggregate error codes. Might need to consider to convert error codes into individual Envoy counters.

> The increase in memory use, as noted in the changes to the tests, seems pretty steep, given that it applies to non-quic configurations as well.
